src/analyzer.rs

```diff
@@ -9,437 +9,503 @@ use std::collections::{BTreeMap, HashMap};
-                // key becomes the struct
-                Some("object") => Some(format!("{}{}", stack, key.to_upper_camel_case())),
-                None => Some(format!("{}{}", stack, key.to_upper_camel_case())),
+                            let vec_value = if let Some(dict_value) = dict_value {
+                                format!("BTreeMap<String, {dict_value}>")
+                            } else {
+                                let structsuffix = key.to_upper_camel_case();
+                                format!("{stack}{structsuffix}")
+                            };
 
-                // leftovers, array of arrays?... need a better way to recurse probably
-                Some(x) => bail!("unknown inner empty dict type {} for {}", x, key),
-            }
-        }
-        "object" => {
-            // cluster test with `failureDomains` uses this spec format
-            Some(format!("{}{}", stack, key.to_upper_camel_case()))
-        }
-        "" => {
-            if s.x_kubernetes_int_or_string.is_some() {
-                Some("IntOrString".into())
-            } else {
-                bail!("unknown empty dict type for {}", key)
+                            Ok((format!("Vec<{}>", vec_value), level))
+                        }
+                        "string" => Ok(("Vec<String>".into(), level)),
+                        "boolean" => Ok(("Vec<bool>".into(), level)),
+                        "date" => Ok((format!("Vec<{}>", extract_date_type(value)?), level)),
+                        "number" => Ok((format!("Vec<{}>", extract_number_type(value)?), level)),
+                        "integer" => Ok((format!("Vec<{}>", extract_integer_type(value)?), level)),
+                        "array" => Ok(self.array_recurse_for_type(s, stack, key, level + 1)?),
+                        unknown => {
+                            bail!("unsupported recursive array type \"{unknown}\" for {key}")
+                        }
+                    };
+                }
+                // maybe fallback to serde_json::Value
+                _ => bail!("only support single schema in array {}", key),
             }
+        } else {
+            bail!("missing items in array type")
         }
-        "boolean" => Some("bool".to_string()),
-        "integer" => Some(extract_integer_type(s)?),
-        // think the type we get is the value type
-        x => Some(x.to_upper_camel_case()), // best guess
-    };
-
-    Ok(dict_key)
+    }
 }
 
-// recurse into an array type to find its nested type
-// this recursion is intialised and ended within a single step of the outer recursion
-fn array_recurse_for_type(
-    value: &JSONSchemaProps,
-    stack: &str,
-    key: &str,
-    level: u8,
-) -> Result<(String, u8)> {
+// ----------------------------------------------------------------------------
+// helpers
+fn is_conditions(value: &JSONSchemaProps) -> bool {
```

Comment:
optional suggestion; you could make this fn return `Option<Schema>` (or whatever inner type `p` is) so you can use `?` to avoid all the `.get() + `is_some()` calls

Reply:
you can see how this plays out in https://github.com/kube-rs/kopium/pull/204/files#diff-f1138a93cb7662493e75aa62afc928d68c742aa12a4d0608f5f845604bce4e54R12-R22 . i feel this is slightly nicer as we have no further need for self and it makes things a little clearer.

Reply:
can't find the suggested changes in #204 😅

Reply:
ah, sorry, i put this comment on the wrong thread 🤦 

actually ended up not doing the suggestion i put in this thread because it was actually slightly awkward